Documentation/proposals/202405-agent-daemonset.md

```diff
@@ -0,0 +1,132 @@
+# DaemonSet deployment for Prometheus Agent
+
+* Owners:
+  * [haanhvu](https://github.com/haanhvu)
+* Related Tickets:
+  * [#5495](https://github.com/prometheus-operator/prometheus-operator/issues/5495)
+* Other docs:
+  * n/a
+
+This proposal is about designing and implementing the deployment of Prometheus Agent as DaemonSet. Currently, Prometheus Agent can only be deployed as StatefulSet, which could be considered as “cluster-wide” strategy, meaning one or several high-availability Prometheus Agents are responsible for scraping metrics of the whole cluster. Though this works well for many use cases, some use cases may indeed prefer a “node-specific” strategy (DaemonSet), where each Prometheus Agent only scrapes the metrics from the targets located on the same node.
+
+
+## 1. Why
+
+When deploying Prometheus Agent in Kubernetes, three of the biggest users’ concerns are: load distribution, scalability, and security.
+
+DaemonSet deployment solves all these three concerns:
+* Load distribution: Each Prometheus Agent pod will only scrape the targets located on the same node. Even though the targets on some nodes may produce more metrics than other nodes, the load distribution would be reliable enough.
+* Automatic scalability: When new nodes are added to the cluster, new Prometheus Agent pods will be automatically added in the nodes that meet user-defined restrictions (if any).
+* Security: Since the scraped targets are local to the Prometheus Agent pod (on the same node), the scope of security problems is reduced to each node.
+
+DaemonSet deployment is especially more suitable for Prometheus Agent than Prometheus, since the storage requirement for the Agent mode is pretty light compared to a fully functional Prometheus server.
+
+This deployment mode has been implemented in [Google Cloud Managed Service for Prometheus (GMP)](https://github.com/GoogleCloudPlatform/prometheus-engine/), so we have an implementation example to learn from. Also, since this deployment has been tested and proved with GMP’s userbase, we can count on that it can solve a large enough number of use cases.
+
+
+## 2. Pitfalls of the current solution
+
+The current (StatefulSet) deployment brings long the corresponding pitfalls:
+* Load management & scalability: Since one or several high-availability Prometheus Agents are responsible for scraping metrics of the whole cluster, users would need to calculate/estimate the load and scalability of the whole cluster to decide on replicas and sharding strategies. Estimating cluster-wide load and scalability is a much harder task than estimating node-wide load and scalability.
+* Security: Similarly, cluster-wide security is a much bigger problem than node-wide security.
```

Comment:
Should we provide examples of how cluster-wide attacks can be avoided with daemonsets?

Reply:
Actually I'm thinking of removing Security from Why and Pitfalls of current solution.

First reason is AFAIU StatefulSet and DaemonSet face different kinds of security issues. We can hardly say one is better than the other. In DaemonSet, Prometheus Agent pod knows all secrets and share them with all other pods (GMP already has a proposal to solve this, but it's not implemented yet). In StatefulSet, the scope of security is larger (at cluster scope) and we have to deal with network issues.

Second reason is security is not the key reason we choose to implement DaemonSet.

You agree to remove it from our consideration?

@simonpasquier @kakkoyun

Reply:
Given the comment above, I'm in favor of removing or at least re-writing the security concerns. I don't think we should be saying that one is better than the other, but acknowledging how they are different.

Reply:
Removed it. @simonpasquier let us know if you have different opinions.